### ISSUE TITLE
SC2183: printf arguments not checking for `*` argument

### ISSUE THREAD

**Issue by stdedos** [koalaman/shellcheck#1184]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2183
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] ~~It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue~~

#### For new checks and feature suggestions
- [x] ~~shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this~~
- [x] I searched through https://github.com/koalaman/shellcheck/issues and #1171 is not related (neither anything referencing 'printf'

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

MAXIMAL=0

for E in A AB ABC ; do
	if [ "${MAXIMAL}" -lt "$(wc -c <<< "${E}")" ] ; then
		MAXIMAL="$(wc -c <<< "${E}")"
	fi
done

for E in A AB ABC ; do
	printf "%*b: %b\\n" "${MAXIMAL}" "${E}" "${E}"
done

```

#### Here's what shellcheck currently says:
```
$ shellcheck myscript
 
Line 12:
        printf "%*b: %b\\n" "${MAXIMAL}" "${E}" "${E}"
               ^-- SC2183: This format string has 2 variables, but is passed 3 arguments.
```

#### Here's what I wanted or expected to see:

(Nothing). printf first string should be checked for `%*X` arguments and +1 per `*` for expected arguments

**Comment by koalaman** [comment]
Kindly fixed by @ngzhian. Thanks!

**Comment by krry** [comment]
Hi y'all. First off, thanks for your diligence. I've learned a lot already from Shellcheck via ALE in Vim.

I've cleared my latest script of all the lint except for this last nagging warning, which references SC2183 here. It's a messy little one-liner that I borrowed from a friend. I don't know how to clean it up, or if the warning is erroneous.

``` sh
printf -v _hr "%*s" "$(tput cols)" && echo -en "${_hr// /$2-$}" && echo -e "\r\033[20C $1 "
```

Here's a screenshot:
<img width="773" alt="screenshot 2019-03-04 09 04 19" src="https://user-images.githubusercontent.com/439967/53741971-2aa75200-3e5d-11e9-8f52-5c0dcdb6d637.png">

**Comment by koalaman** [comment]
@krry The issue here is that `%*s` takes two parameter: a length and a string to pad to that length:

    $ printf '%*s' 16 "hello world"
         hello world
    ^------16------^

Your snippet leaves out the second argument, so Bash just inserts a type specific default: the empty string. If you just want N spaces, that happens to work out, but it's arguably better to be explicit about it:

    printf -v _hr "%*s" "$(tput cols)" ""

**Comment by stdedos** [comment]
Since you are not posting what is the script supposed to do, why do commands need to be nested like that etc, I will assume I can pretty-much do whatever to the snippet.

----

It is easy to see what is the issue, if you try to break the chaining:
```bash
#!/bin/bash

printf -v _hr "%*s" "$(tput cols)"
#             ^-- SC2183: This format string has 2 variables, but is passed 1 arguments.
echo -en "${_hr// /${2-$}}"
echo -e "\r\033[20C $1 "

```

So, you are trying to do "some" kind of magic, that involves the width of your terminal - some kind of "measurement" specifically (I am familiar with some `printf` "snippets", and they all look hacky to the untrained eye). What `shellcheck` is telling you, is that you asked `printf` to do "something" with 2 arguments (`%*`: min arg width, [...]`s`: of-type: string), and you gave one (min arg width)

Maybe your extra code gives some hints about what `printf` (and the rest of the `echo`s) is/are supposed to do. Warning (could be) easy to fix, and it might be better to do so: you will then be explicit.

You might also understand better the job the tool is performing 😄

**Comment by krry** [comment]
Aha, thank you @koalaman and @stdedos.

I didn't understand that printf could output to a variable, nor that it was looking for two parameters with the `*` in the format.

This snippet has been tough for me to decipher, and it still feels like a hack since I'm using printf and echo side-by-side. The goal is to print terminal-width rows of a character or characters (passed optionally as the second argument) with a message embedded in the center (passed optionally as the first argument). And it's working. I just don't understand why.

Oh, and there's no need for it to be a one-liner if I wrap it in a function.

```bash
#!/bin/bash
hr_msg ()  {
    printf -v _hr "%*s" "$(tput cols)" ""
    echo -en "${_hr// /${2--}}"
    echo -e "\r\033[$((($(tput cols)-${#1}-2)/2))C ${1-hi} "
}
hr_msg "Look, I'm floating in stars" "*"
hr_msg "What am I, buried treasure?" ".,.,.:.;.:.;"
```

I also want to translate it for fish shell. And get the message to float in the middle of a multi-line output.

**Comment by stdedos** [comment]
tl;dr: keep `printf -v _hr "%*s" "$(tput cols)" ""`

----

> This snippet has been tough for me to decipher, and it still feels like a hack since I'm using printf and echo side-by-side. 

Nooooot really 😝. It's just that enough of us, are hardcoded to `echo`, but then, when we need to do something "fancy", we use `printf`.

And then back to `echo` 😉 

It's not hard to convert everything to printf, but then, it's a question of "how much effort" are you willing to spend to make it backwards-compatible? and yada-yada.

Writing it "once" is better than going back and forth changing from one implementation to another.

----

On a sidenote, function looks weird on a multi-line delimiter:

```bash
user@host [2019-03-13 20:42:34]:~# hr_msg "Look, I'm floating in stars" "*"
********************************************* Look, I'm floating in stars **********************************************
user@host [2019-03-13 20:42:36]:~# hr_msg "What am I, buried treasure?" ".,.,.:.;.:.;"
.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;. What am I, buried treasure? .,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;.,.,.:.;.:.;
user@host [2019-03-13 20:42:39]:~#
```

but I guess that's beside the point of the warning anyway.
